Add prometheus metrics exporter

Ranhiru opened the pull request:
Opening an early Draft PR to get some feedback around a Prometheus exporter using [Prometheus Python Client](https://github.com/prometheus/client_python)

Labels are following the [Metric and Label naming guide ](https://prometheus.io/docs/practices/naming/)

The metrics output looks like below.

```
# HELP inverter_device_type_code_info Device Type Code
# TYPE inverter_device_type_code_info gauge
inverter_device_type_code_info{device_type_code="SG5K-D"} 1.0
# HELP inverter_run_state Run State
# TYPE inverter_run_state gauge
inverter_run_state{inverter_run_state="ON"} 1.0
inverter_run_state{inverter_run_state="OFF"} 0.0
# HELP inverter_daily_power_yield_watt_hours_total Daily Power Yields
# TYPE inverter_daily_power_yield_watt_hours_total gauge
inverter_daily_power_yield_watt_hours_total 1800.0
# HELP inverter_total_power_yield_watt_hours_total Total Power Yields
# TYPE inverter_total_power_yield_watt_hours_total gauge
inverter_total_power_yield_watt_hours_total 70100.0
# HELP inverter_internal_temperature_celsius Internal Temperature
# TYPE inverter_internal_temperature_celsius gauge
inverter_internal_temperature_celsius 34.0
# HELP inverter_phase_a_voltage_volts Phase A Voltage
# TYPE inverter_phase_a_voltage_volts gauge
inverter_phase_a_voltage_volts 235.2
# HELP inverter_active_power_watts_total Total Active Power
# TYPE inverter_active_power_watts_total gauge
inverter_active_power_watts_total 1482.0
# HELP inverter_work_state_info Work State
# TYPE inverter_work_state_info gauge
inverter_work_state_info{work_state="Run"} 1.0
# HELP inverter_meter_power_watts_total Meter Power
# TYPE inverter_meter_power_watts_total gauge
inverter_meter_power_watts_total -677.0
```

I haven't added all the metrics yet but wondering if this is ok or do I need to maybe push the configuration in to the `config.yaml` ?

michbeck100 commented:
Thanks for your efforts, but there is already a `/metrics` endpoint in the webserver exporter (https://github.com/bohdan-s/SunGather/commit/bca665ade308027a5ce4b143e6d99a073a262727). Maybe you could extend this one? I don’t think we need an extra exporter.

Ranhiru commented:
Hey @michbeck100. Thank you for your input. 

Hopefully I can convince you to have a separate Prometheus exporter 🙂 

1. The existing `/metrics` endpoint is currently not working. This is due to it incorrectly exporting multiple string values (inverter name, work state) [that's not supported ](https://github.com/prometheus/prometheus/issues/2227)
2. The naming convention for the currently exported metrics does not follow [naming best practices](https://prometheus.io/docs/practices/naming/#metric-names)
3. The values that are exported are not in[ their base units](https://prometheus.io/docs/practices/naming/#base-units)
4. There's no ability currently to configure which metrics are exported 
5. I believe the current metrics are exported incorrectly in regards to the use of labels (Please allow me to say that I'm very new Prometheus, so it's very possible I'm wrong on this one)

For example one of the currently exported metrics is shown below
```
daily_power_yields{address="5003", unit="kWh"} 17.6
```

The use of `address` and `unit` as labels I believe is incorrect as labels are used to create additional dimensions for the same metric. Please see below an example I've copied from this [web page](https://blog.pvincent.io/2017/12/prometheus-blog-series-part-1-metrics-and-labels/)

```
# Request counter for the User Directory service
http_requests_total{service="users-directory"}

# Request counter for the Billing History Service
http_requests_total{service="billing-history"}
```

The address (5003) or unit (kWh) is not going to change for the life time of the metrics gathering.
6. Finally and most importantly I've created a new export so as to not disrupt the current usage of the `/metrics` endpoint.

michbeck100 commented:
I still don’t see the need for a dedicated exporter. You can reuse and change/fix the current `/metrics` endpoint. It makes no sense to have two endpoints for prometheus. I like all the other things you pointed out. I’ve implemented the endpoint in the webserver exporter because it’s just a different format not worth a whole exporter.

bohdan-s commented:
It looks like we could add this to the existing /metrics endpoint with a few extra settings in the config file? Would there a be use case for both styles to run at the same time though?

Ranhiru commented:
There would be no need to have two exporters, that I fully agree with. However I'm sure there's active usage of the current `/metrics` endpoint by others and didn't want to disrupt that. My idea was to run things side by side and deprecate the old endpoint

michbeck100 commented:
@Ranhiru just extend/replace the existing one. I think the endpoint should still be /metrics. Since it will be a new release version we can replace it and people can switch to the new version once they adopted to the new format.

Ranhiru commented:
Hey @michbeck100, apologies for not getting back to you sooner, it's been a busy week.

I'm still hoping I can convince you or @bohdan-s to make metrics its own exporter. I'd say that if anything, the HTML and JSON content should have been served through the same exporter with different `Content-Type`.

Although Prometheus style of metrics is different from HTML or JSON, I understand that it's still plain text served through a web endpoint. (`/metrics`) I'll see what's the best way to get the metrics integrated to the web exporter instead of creating a new exporter.

I have a question around the current configuration [I've added here](https://github.com/bohdan-s/SunGather/pull/73/files#diff-abc97811d4954606a27957911b8465f6b206919f9f57dc93c8835084c61115b3R158). I've added configuration for each of the fields that's returned from my inverter but I'm assuming not all inverters will support all the fields. Do I add some example configuration and leave it up to the user to fill in the rest ?

Ranhiru commented:
Not the nicest looking dashboard but I could get this working in Grafana Cloud

<img width="1321" alt="Screen Shot 2022-08-15 at 4 39 22 pm" src="https://user-images.githubusercontent.com/2367456/184588161-08df858c-3070-4e39-adbd-f8c5230f4999.png">

michbeck100 commented on SunGather/prometheus/metrics_collector.py:
```diff
@@ -0,0 +1,92 @@
+        metric_name = metric_config['metric_name']
+        metric_type = metric_config['type']
+
+        logging.info(f"Prometheus: Initialising metric {metric_name} as type {metric_type}")
```
Maybe rather `logging.debug`?

Ranhiru replied:
Cool, I'll change that

michbeck100 commented:
> Do I add some example configuration and leave it up to the user to fill in the rest ?

I think that's ok for now. But for the future we should try to implement an automatism that exports all registers without extra config.

michbeck100 commented:
Basically I like the implementation. What I don't like is the rather complex configuration. That's what I like about the current implementation which just exports all registers. Would be great to have this ability, too. Maybe decide by `datatype` which metric type to use and create the description from register name. And reuse the unit. 

What do you think?

Ranhiru commented:
@michbeck100 Fair enough. The configuration is definitely complex.

How about instead of having all of the current configuration in the config.yaml, we leave only the most commonly used metrics that the influx db exporter currently exports, and move the rest as examples to the README or a Wiki ?

In that way, we let the end user decide how complex they want the configuration to be

Ranhiru commented:
Anything else that needs to be done to merge this?
